artichoke-backend/src/convert/implicit.rs

```diff
@@ -483,3 +486,213 @@ pub unsafe fn implicitly_convert_to_nilable_string<'a>(
+///   - If `value.to_str` raises an exception, propagate that exception.
+///   - If `value.to_str` returns a Ruby string, return the inner byte slice.
+///   - If `value.to_str` returs any other type, return a [`TypeError`].
+/// - If the given value does not respond to the `:to_str` method, return a
+///   [`TypeError`].
+///
+/// [`Symbol`]s are not coerced to byte slice by this implicit conversion.
+///
+/// # Examples
+///
+/// ```
+/// # use artichoke_backend::prelude::*;
+/// # use artichoke_backend::convert::implicitly_convert_to_spinoso_string;
+/// # use artichoke_backend::value::Value;
+/// # fn example() -> Result<(), Error> {
+/// let mut interp = artichoke_backend::interpreter()?;
+/// // successful conversions
+/// let mut string = interp.try_convert_mut("artichoke")?;
+/// let mut a = interp.eval(b"class A; def to_str; 'spinoso'; end; end; A.new")?;
+///
+/// # unsafe {
+/// assert!(matches!(implicitly_convert_to_spinoso_string(&mut interp, &mut string), Ok(s) if *s == b"artichoke"[..]));
+/// assert!(matches!(implicitly_convert_to_spinoso_string(&mut interp, &mut a), Ok(s) if *s == b"spinoso"[..]));
+///
+/// // failed conversions
+/// let mut nil = Value::nil();
+/// let mut b = interp.eval(b"class B; end; B.new")?;
+/// let mut c = interp.eval(b"class C; def to_str; nil; end; end; C.new")?;
+/// let mut d = interp.eval(b"class D; def to_str; 17; end; end; D.new")?;
+/// let mut e = interp.eval(b"class E; def to_str; :artichoke; end; end; E.new")?;
+/// let mut f = interp.eval(b"class F; def to_str; raise ArgumentError, 'not a string'; end; end; F.new")?;
+///
+/// assert!(implicitly_convert_to_spinoso_string(&mut interp, &mut nil).is_err());
+/// assert!(implicitly_convert_to_spinoso_string(&mut interp, &mut b).is_err());
+/// assert!(implicitly_convert_to_spinoso_string(&mut interp, &mut c).is_err());
+/// assert!(implicitly_convert_to_spinoso_string(&mut interp, &mut d).is_err());
+/// assert!(implicitly_convert_to_spinoso_string(&mut interp, &mut e).is_err());
+/// assert!(implicitly_convert_to_spinoso_string(&mut interp, &mut f).is_err());
+/// # }
+/// # Ok(())
+/// # }
+/// # example().unwrap();
+/// ```
+///
+/// # Errors
+///
+/// This function returns an error if:
+///
+/// - The given value is `nil`.
+/// - The given value is not a string and does not respond to `:to_str`.
+/// - The given value is not a string and returns a non-string value from its
+///   `:to_str` method.
+/// - The given value is not a string and raises an error in its `:to_str`
+///   method.
+///
+/// # Safety
+///
+/// Callers must ensure that `value` does not outlive the given interpreter.
+///
+/// If a garbage collection can possibly run between calling this function and
+/// using the returned slice, callers should convert the slice to an owned byte
+/// vec.
+///
+/// [`Symbol`]: crate::extn::core::symbol::Symbol
+pub unsafe fn implicitly_convert_to_spinoso_string<'a>(
```

Comment:
We have both `implicitly_convert_to_string` and `implicitly_convert_to_nilable_string`. If you find a need for implicitly converting to a nilable spinoso string, it is just a few lines to add that function. Since I don't need it here, I'd prefer to hold off on adding it.

https://github.com/artichoke/artichoke/blob/9a979d9f57ad2afe4b7d312e91a07e15aee93d28/artichoke-backend/src/convert/implicit.rs#L475-L485

Reply:
> the switch logic needs to handle 3 types in this above case

One thing that I think is a constant across all of the type conversions is that only one argument shape gets the implicit conversion treatment. So a method will either implicitly convert to string OR int, but not both.

That implicit conversion needs to go last in the chain so we can propagate the `TypeError`s.

Reply:
> Since I don't need it here, I'd prefer to hold off on adding it.

Yes good call!